src/webgpu/shader/types.ts

```diff
@@ -261,51 +299,64 @@ export function* generateTypes({
 
   // Array types
   if (containerType === 'array') {
-    // Buffer affective binding size must be a multiple of 4. Adjust array length as needed.
-    let arrayLength = kDefaultArrayLength;
-    if (
-      addressSpace === 'storage' &&
-      scalarInfo.layout !== undefined &&
-      scalarInfo.layout.alignment % 4 !== 0
-    ) {
-      arrayLength = align(arrayLength, 4);
+    let arrayElemType: string = scalarType;
+    let arrayElementCount: number = kDefaultArrayLength;
+    let supportsAtomics = scalarInfo.supportsAtomics;
+    let layout: undefined | AlignmentAndSize = undefined;
+    if (scalarInfo.layout) {
+      // Compute the layout of the array type.
+      // Adjust the array element count or element type as needed.
+      if (addressSpace === 'uniform') {
+        // Use a vec4 of the scalar type, to achieve a 16 byte alignment without internal padding.
+        // This works for 4-byte scalar types, and does not work for f16.
+        // It is the caller's responsibility to filter out the f16 case.
+        assert(!isAtomic, 'the uniform case is making vec4 of scalar, which cannot handle atomics');
+        arrayElemType = `vec4<${baseType}>`;
+        supportsAtomics = false;
+        const arrayElemLayout = vectorLayout('vec4', baseType) as AlignmentAndSize;
+        // assert(arrayElemLayout.alignment % 16 === 0); // Callers responsibility to avoid
+        arrayElementCount = align(arrayElementCount, 4) / 4;
+        const arrayByteSize = arrayElementCount * arrayElemLayout.size;
+        layout = { alignment: arrayElemLayout.alignment, size: arrayByteSize };
+      } else {
+        // The ordinary case.  Use scalarType as the element type.
+        const stride = arrayStride(scalarInfo.layout);
+        let arrayByteSize = arrayElementCount * stride;
+        if (addressSpace === 'storage') {
+          // The buffer effective binding size must be a multiple of 4.
+          // Adjust the array element count as needed.
+          while (arrayByteSize % 4 > 0) {
+            arrayElementCount++;
+            arrayByteSize = arrayElementCount * stride;
+          }
+        }
+        layout = { alignment: scalarInfo.layout.alignment, size: arrayByteSize };
+      }
     }
 
     const arrayTypeInfo = {
       elementBaseType: `${baseType}`,
-      arrayLength,
-      layout: scalarInfo.layout
-        ? {
-            alignment: scalarInfo.layout.alignment,
-            size:
-              addressSpace === 'uniform'
-                ? // Uniform storage class must have array elements aligned to 16.
-                  arrayLength *
-                  arrayStride({
-                    ...scalarInfo.layout,
-                    alignment: 16,
-                  })
-                : arrayLength * arrayStride(scalarInfo.layout),
-          }
-        : undefined,
+      arrayLength: arrayElementCount,
+      layout: scalarInfo.layout ? layout : undefined,
```

Comment:
nit: I think this could equivalently just be:
```suggestion
      layout,
```